### PR TITLE
Feat: Earned badge and ranked popup; Closes #551

### LINKED ISSUE #551: Popup of Badge after Earning
Badges should be celebrated more. After a student earns a badge, a window should popup saying congratulations you earned x badge along with the details of that badge and how many bonus XP they earned with the badge. 

This would make badges a more powerful reinforcer and create more of an interest in earning badges. 

Khan Academy approaches the earning of badges in this way. A victory sound is also played when a badge is earnt.

### PR DESCRIPTION
**Pull request by 9921622**
Please ensure you are familiar with our Pull Request Description expectations described here: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/
### What?
Created a popup everytime a badge or rank is earned. User's can dismiss the popup by clicking on "dismiss". Refreshing or clearing the popup by clicking on the greyed out background will trigger the popup again when page is reset.

### Why?
See #551

#### Badge
Created an ajax request in ```javascript.js``` linked to ```base.html.```
On webpage startup calls `courses:ajax_on_show_badge_popup` 

The ajax request calls a view that returns a `JsonResponse` with html the javascript can load into the page.
The html snippet has all the recent badges the user has gotten, using unmarked badge notifications.

#### Rank
When user has their quest approved a new notification is sent out if their new xp reaches a new rank.

When webpage is loaded an ajax request in ```javascript.js```( linked to ```base.html.```) is called to `courses:ajax_on_show_ranked_popup`.

The ajax request then returns a `JsonResponse` with html, then with client side code can load into the page.
The html snippet has `earned_rank`, the new rank the user has achieved, and `next_rank`, the next rank the user can achieve.
The request gets `earned_rank` by filtering the users unread notifications for the new "promoted" notification (see screenshots).
 
### Testing?
Created ajax tests
`BadgeAjaxTests` tests for `Ajax_OnShowBadgePopup` and `Ajax_OnCloseBadgePopup`
`AjaxRankPopupTests` tests for `Ajax_OnShowRankPopup` and `Ajax_OnCloseRankPopup`
Created notification testing for `complete` and `approved` views for the new notification type
Created notification testing for `post_save_reciever` for badge assertions

Did not add testing for `QuestSubmission.does_not_grant_xp` as `complete` and `approved` views use `submission.mark_approved()`. Which automatically sets `does_not_grant_xp` to `False`

### Screenshots (if front end is affected)

Video of both popups showing up and being dismissed

https://github.com/user-attachments/assets/83bcb51f-8061-427b-bee1-26e8ada35255

#### Badges
Badge popup
image
image

#### Ranks
new notification for rank up

Ranked popup
image

ranked popup with a related map
image

New notification to indicate rank. Recieved when user's xp goes over the threshold when getting a new badge or when quest is approved
image

### Anything Else?
+ Added a new decorator. `xml_http_request_required` which checks if request is an ajax request
+ Fixed a bug where calling `ajaxNotificationsBadge()` with notifications then calling it again does not clear `.notification-badge`

cant seem to add a ranked object to a map if its already related to another map. Is that a bug?

### Review request
@tylerecouture 

## Summary by CodeRabbit

- **New Features**
  - Added a notification feature for rank upgrades.
  - Introduced badge and rank popups that display congratulatory messages dynamically.
  - Added AJAX functionality for showing and closing badge and rank popups.
  - New decorators to enhance AJAX request handling.

- **Bug Fixes**
  - Ensured correct creation and display of notifications for badge assertions and rank changes.

- **Tests**
  - Added comprehensive tests for badge and rank popup notifications.
  - Enhanced tests for quest submissions and XP granting logic.

- **Chores**
  - Updated templates to include AJAX functions for badge and rank notifications.

### CODE REVIEW

**Comment by 9921622** [comment]
@tylerecouture do you think my approach of using unread notifications as a base for new badges is good? I want to know before I start using the same method for the new rank popups

**Comment by tylerecouture** [comment]
> @tylerecouture do you think my approach of using unread notifications as a base for new badges is good? I want to know before I start using the same method for the new rank popups

neat idea. I'm not sure! Let me think about it.

**Comment by tylerecouture** [comment]
Could this be simplified a bit, if dismissing the popup also marked those specific badge notifications as read?

**Comment by 9921622** [comment]
@tylerecouture review?

**Comment by 9921622** [comment]
wait forgot to add tests for the `notify_rank_up` in `badges.models.post_save_receiver`

**Comment by 9921622** [comment]
@tylerecouture review

File: src/badges/tests/test_views.py [hunk 1]
```diff
@@ -414,3 +418,87 @@
         # Get Delete view and assert header is correct
         request = self.client.get(reverse('badges:badge_type_delete', args=[self.badge_type.id]))
         self.assertContains(request, 'Delete CustomBadge Type')
+
+
+class BadgeAjaxTests(ViewTestUtilsMixin, TenantTestCase):
+    """ test case for
+    + ajax/on_show_badge_popup/
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/tests/test_views.py:425]: Why shouldn't we use hardcoded passwords in tests?

**Review by tylerecouture — COMMENTED** [review]
Can you show me what this looks like with only 1 badge, also 3 and 4 different badges (you show 2 here):
image

**Review by tylerecouture — COMMENTED** [review]
One feature that is kind of hidden is that Ranks have their own maps.  Are you able also note that "You have gained access to a new Quest Map by earning this rank" or something like that, with a link to the map, IF a map existing at that rank?

Note that if this would be easier to do with #1631, then hold off and wait cus that one's acceptance is imminent!

**Comment by 9921622** [comment]
> Can you show me what this looks like with only 1 badge, also 3 and 4 different badges (you show 2 here): image
> 
> Also, what happens if you earn a new Rank AND a new badge? Can you screen record?

Theres actually a screenrecord attachment in "Screenshots" im not sure why github isnt formatting it, But its there as a link

File: src/badges/templates/badges/snippets/badge_notification_popup.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      <div class="modal-content">
+        <div class="modal-header">
+          <h5 class="modal-title" id="newBadgeLabel">
+            {% if badge_total == 1 %}
+            Congratulations! You've gained a badge!
+            {% else %}
+            Congratulations! You have earned {{ badge_total }} new {{ badge_name }}s!
+            {% endif %}
+          </h5>
+          <button type="button" class="close" data-dismiss="modal" aria-label="Close">
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/templates/badges/snippets/badge_notification_popup.html:13]: Dismiss button checks off notifications and closes the popup. 
'X' button only closes the popup (as well as clicking outside of the popup)

I think i just remove the 'X'

File: src/courses/templates/courses/snippets/rank_notification_popup.html [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      <div class="modal-header">
+        <h5 class="modal-title" id="newRankLabel">New Rank!</h5>
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+        </button>
+      </div>
+      <div class="modal-body text-center">
+
+
+        <div class="row align-items-center mb-3">
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/courses/templates/courses/snippets/rank_notification_popup.html:14]: mb is margin bottom. Doesnt exist in bootstrap 3. removed

**Comment by 9921622** [comment]
@tylerecouture review

**Review by tylerecouture — APPROVED** [review]
Some minor issues that it's probably easier for me to clean up myself =D

**Pull request merged**